File: Schemes.Repository/CustomerRepository.cs
using Microsoft.EntityFrameworkCore;
using Schemes.Models;
using Schemes.Repository.Migrations;
using Schemes.ViewModels;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading.Tasks;
using static Schemes.ViewModels.Enums;

namespace Schemes.Repository
{
    public class CustomerRepository
    {
        private readonly SchemesContext _dbContext;
        public CustomerRepository(SchemesContext DbContext)
        {
            _dbContext = DbContext;
        }
        public List<SchemeDetails> GetSchemesList(string name)
        {
            List<SchemeDetails> schemeDetailsList =new List<SchemeDetails>();
             var details   = _dbContext.SchemesDetails.Where(s=>s.AvaliableFor.Contains(name) &&s.IsActive==true).ToList();
            foreach (var scheme in details)
            {
                SchemeDetails schemeDetails=new SchemeDetails();
                schemeDetails.SchemesDetailID = scheme.SchemesDetailID;
                schemeDetails.Area = scheme.Area;
                schemeDetails.ApplyAndLink = scheme.ApplyAndLink;
                schemeDetails.NameOftheScheme = scheme.NameOftheScheme;
                schemeDetails.Benefits = scheme.Benefits;
                schemeDetails.EligibilityCriteria = scheme.EligibilityCriteria;
                schemeDetails.Description = scheme.Description;
                schemeDetails.DocumentsRequired = scheme.DocumentsRequired;
                schemeDetails.AvailableFor = scheme.AvaliableFor;

                schemeDetailsList.Add(schemeDetails);
            }
            return schemeDetailsList;
        }
        public List<SchemeDetails>? GetSchemesListByLangCode(string name, string langCode)
        {
            List<SchemeDetails>? schemeDetailsList = new List<SchemeDetails>();
            var schemeDetailsFromDB = _dbContext.SchemesDetails.Where(s => s.AvaliableFor.Contains(name)&&s.IsActive == true).ToList();
            if (schemeDetailsFromDB != null)
            {
                var details = _dbContext.MultilingualSchemesData.Where(s => s.AvaliableFor.Contains(name) && s.LangCode == langCode && s.IsActive == true).ToList();
                if (details != null)
                {
                    foreach (var scheme in details)
                    {
                        SchemeDetails schemeDetails = new SchemeDetails();
                        schemeDetails.SchemesDetailID = scheme.SchemesDetailID;
                        schemeDetails.Area = scheme.Area;
                        schemeDetails.ApplyAndLink = scheme.ApplyAndLink;
                        schemeDetails.NameOftheScheme = scheme.NameOftheScheme;
                        schemeDetails.Benefits = scheme.Benefits;
                        schemeDetails.EligibilityCriteria = scheme.EligibilityCriteria;
                        schemeDetails.Description = scheme.Description;
                        schemeDetails.DocumentsRequired = scheme.DocumentsRequired;

                        schemeDetailsList.Add(schemeDetails);
                    }
                }
                else
                {
                    foreach (var scheme in schemeDetailsFromDB)
                    {
                        SchemeDetails schemeDetails = new SchemeDetails();
                        schemeDetails.SchemesDetailID = scheme.SchemesDetailID;
                        schemeDetails.Area = scheme.Area;
                        schemeDetails.ApplyAndLink = scheme.ApplyAndLink;
                        schemeDetails.NameOftheScheme = scheme.NameOftheScheme;
                        schemeDetails.Benefits = scheme.Benefits;
                        schemeDetails.EligibilityCriteria = scheme.EligibilityCriteria;
                        schemeDetails.Description = scheme.Description;
                        schemeDetails.DocumentsRequired = scheme.DocumentsRequired;

                        schemeDetailsList.Add(schemeDetails);
                    }
                }
            }
            return schemeDetailsList;
        }
        public List<SchemeDetails> GetAllSchemesList()
        {
            List<SchemeDetails> schemeDetailsList = new List<SchemeDetails>();
            var details = _dbContext.SchemesDetails.Where(s=>s.IsActive==true).ToList();
            foreach (var scheme in details)
            {
                SchemeDetails schemeDetails = new SchemeDetails();
                schemeDetails.SchemesDetailID = scheme.SchemesDetailID;
                schemeDetails.Area = scheme.Area;
                schemeDetails.ApplyAndLink = scheme.ApplyAndLink;
                schemeDetails.NameOftheScheme = scheme.NameOftheScheme;
                schemeDetails.Benefits = scheme.Benefits;
                schemeDetails.EligibilityCriteria = scheme.EligibilityCriteria;
                schemeDetails.Description = scheme.Description;
                schemeDetails.DocumentsRequired = scheme.DocumentsRequired;
                schemeDetails.IsActive = scheme.IsActive;
                schemeDetails.AvailableFor = scheme.AvaliableFor;


                schemeDetailsList.Add(schemeDetails);
            }
            return schemeDetailsList;
        }
        public int AddNewScheme(SchemeDetails scheme)
        {
            int schemeID = 0;
            Models.SchemesDetails schemeDetails = new SchemesDetails();
                schemeDetails.Area = scheme.Area;
                schemeDetails.ApplyAndLink = scheme.ApplyAndLink;
                schemeDetails.NameOftheScheme = scheme.NameOftheScheme;
                schemeDetails.Benefits = scheme.Benefits;
                schemeDetails.EligibilityCriteria = scheme.EligibilityCriteria;
                schemeDetails.Description = scheme.Description;
                schemeDetails.DocumentsRequired = scheme.DocumentsRequired;
                schemeDetails.AvaliableFor = scheme.AvailableFor;
                schemeDetails.IsActive = true;
            schemeDetails.InsertedDate = DateTime.Now;
            schemeDetails.InsertedBy = "Admin";
               var details= _dbContext.SchemesDetails.Add(schemeDetails).Entity;
                _dbContext.SaveChanges();

            List<SMSDetails> sMSDetails = new List<SMSDetails>();
            var customersList = GetCustomersList();
            var text = $"Fineasee:Dear Customer, New Scheme {scheme.NameOftheScheme} is available for: {scheme.AvailableFor} on our website";
            foreach (var customer in customersList)
            {
                new LoginRepository(_dbContext).SendEmailToUser(customer.EmailId, text, false);
                sMSDetails.Add(new SMSDetails()
                {
                    phone_number = customer.PhoneNo,
                    text_message = text
                });
            }
            new LoginRepository(_dbContext).SendOTPAsync(sMSDetails);
            MultilingualSchemesDataVM schemeData = new MultilingualSchemesDataVM();
            schemeData.Area = scheme.Area;
            schemeData.ApplyAndLink = scheme.ApplyAndLink;
            schemeData.NameOftheScheme = scheme.NameOftheScheme;
            schemeData.Benefits = scheme.Benefits;
            schemeData.EligibilityCriteria = scheme.EligibilityCriteria;
            schemeData.Description = scheme.Description;
            schemeData.DocumentsRequired = scheme.DocumentsRequired;
            schemeData.AvaliableFor = scheme.AvailableFor;
            AddNewSchemeByLangCode(schemeData, details.SchemesDetailID, "en");
            schemeID = details.SchemesDetailID;
            return schemeID;
        }
        public bool AddNewSchemeByLangCode(MultilingualSchemesDataVM scheme,int schemeID,string LangCode)
        {
            Models.MultilingualSchemesData schemeDetails = new MultilingualSchemesData();
            schemeDetails.Area = scheme.Area;
            schemeDetails.ApplyAndLink = scheme.ApplyAndLink;
            schemeDetails.NameOftheScheme = scheme.NameOftheScheme;
            schemeDetails.Benefits = scheme.Benefits;
            schemeDetails.EligibilityCriteria = scheme.EligibilityCriteria;
            schemeDetails.Description = scheme.Description;
            schemeDetails.DocumentsRequired = scheme.DocumentsRequired;
            schemeDetails.AvaliableFor = scheme.AvaliableFor;
            schemeDetails.LangCode = LangCode;
            schemeDetails.SchemesDetailID = schemeID;
            schemeDetails.IsActive = true;
            schemeDetails.InsertedDate = DateTime.Now;
            schemeDetails.InsertedBy = "Admin";
            _dbContext.MultilingualSchemesData.Add(schemeDetails);
            _dbContext.SaveChanges();
            return true;
        }
        public bool UpdateScheme(SchemeDetails scheme)
        {

            var details = _dbContext.SchemesDetails.Where(s => s.SchemesDetailID == scheme.SchemesDetailID).FirstOrDefault();
            if (details != null)
            {
                details.Area = scheme.Area;
                if(scheme.ApplyAndLink !=null ) {
                  details.ApplyAndLink=  ReplaceUrlsWithLinks(scheme.ApplyAndLink);
                }
                details.NameOftheScheme = scheme.NameOftheScheme;
                details.Benefits = scheme.Benefits;
                details.EligibilityCriteria = scheme.EligibilityCriteria;
                details.Description = scheme.Description;
                details.DocumentsRequired = scheme.DocumentsRequired;
                details.AvaliableFor = scheme.AvailableFor;
                details.UpdatedDate = DateTime.Now;
                details.UpdatedBy = "Admin";
                _dbContext.SaveChanges();
                return true;
            }
            return false;
        }
        public bool UpdateSchemeStatus(int schemeID)
        {

            var details = _dbContext.SchemesDetails.Where(s => s.SchemesDetailID == schemeID).FirstOrDefault();
            if (details != null)
            {
                if ((bool)details.IsActive)
                {
                    details.IsActive = false;
                }
                else
                {
                    details.IsActive = true;
                }
                _dbContext.SaveChanges();

                return true;
            }
            return false;
        }
        public bool UpdateCustomerStatus(int customerID)
        {

            var details = _dbContext.Customers.Where(s => s.CustomerId == customerID).FirstOrDefault();
            if (details != null)
            {
                
                    details.CustomerStatus = (int)CustomerStatus.InActive;
                _dbContext.SaveChanges();

                return true;
            }
            return false;
        }

        public List<CustomerDetails> GetCustomersList()
        {
            List<CustomerDetails> customersList = new List<CustomerDetails>();
            customersList = _dbContext.Customers.Select(s=> new CustomerDetails()
            {
                CustomerId=s.CustomerId,
                DateOfBirth=s.DateOfBirth,
                EmailId=s.EmailId,
                FirstName=s.FirstName,
                PhoneNo=s.PhoneNo,
                LastName=s.LastName,
                CustomerStatus=(CustomerStatus)s.CustomerStatus
            }).ToList();
            
            return customersList;
        }
        public SchemeDetails GetSchemesDetails(int schemeiD)
        {
            SchemeDetails schemeDetails = new SchemeDetails();
            var scheme = _dbContext.SchemesDetails.Where(s => s.SchemesDetailID== schemeiD).FirstOrDefault();
            if (scheme != null)
            {
                schemeDetails.SchemesDetailID = scheme.SchemesDetailID;
                schemeDetails.Area = scheme.Area;
                schemeDetails.ApplyAndLink = scheme.ApplyAndLink;
                schemeDetails.NameOftheScheme = scheme.NameOftheScheme;
                schemeDetails.Benefits = scheme.Benefits;
                schemeDetails.EligibilityCriteria = scheme.EligibilityCriteria;
                schemeDetails.Description = scheme.Description;
                schemeDetails.DocumentsRequired = scheme.DocumentsRequired;
                schemeDetails.AvailableFor = scheme.AvaliableFor;
            }
            
            return schemeDetails;
        }
        public LoanDetails GetLoanDetails(string bankName)
        {
            LoanDetails loanDetails = new LoanDetails();
            var loanInterestDetails = _dbContext.LoanInterestDetails.Where(s => s.BankName.Contains(bankName)).FirstOrDefault();
            if (loanInterestDetails != null)
            {
                loanDetails.LoanType = loanInterestDetails.LoanType;
                loanDetails.BankName =loanInterestDetails.BankName;
                loanDetails.MaxloanAmount = loanInterestDetails.MaxloanAmount;
                loanDetails.MinLoanAmount = loanInterestDetails.MinLoanAmount;
                loanDetails.MaxInterestRate = loanInterestDetails.MaxInterestRate;
                loanDetails.MinInterestRate = loanInterestDetails.MinInterestRate;

            }

            return loanDetails;
        }
        public List<LoanDetails> GetLoanDetailsList()
        {
            List<LoanDetails> list = new List<LoanDetails>();
            var loanInterestDetailsList = _dbContext.LoanInterestDetails.ToList();
            if (loanInterestDetailsList.Count >0)
            {
                foreach (var loanInterestDetails in loanInterestDetailsList)
                {
                    LoanDetails loanDetails = new LoanDetails();
                    loanDetails.LoanType = loanInterestDetails.LoanType;
                    loanDetails.LoanInterestDetailID = loanInterestDetails.LoanInterestDetailID;
                    loanDetails.BankName = loanInterestDetails.BankName;
                    loanDetails.MaxloanAmount = loanInterestDetails.MaxloanAmount;
                    loanDetails.MinLoanAmount = loanInterestDetails.MinLoanAmount;
                    loanDetails.MaxInterestRate = loanInterestDetails.MaxInterestRate;
                    loanDetails.MinInterestRate = loanInterestDetails.MinInterestRate;
                    list.Add(loanDetails);
                }

            }

            return list;
        }
        public bool AddLoanDetails(LoanDetails loanInterestDetails)
        {
            LoanInterestDetails loanDetails = new LoanInterestDetails();
            if (loanInterestDetails != null)
            {
                loanDetails.LoanType = loanInterestDetails.LoanType;
                loanDetails.BankName = loanInterestDetails.BankName;
                loanDetails.MaxloanAmount = loanInterestDetails.MaxloanAmount;
                loanDetails.MinLoanAmount = loanInterestDetails.MinLoanAmount;
                loanDetails.MaxInterestRate = loanInterestDetails.MaxInterestRate;
                loanDetails.MinInterestRate = loanInterestDetails.MinInterestRate;
                loanDetails.InsertedBy = "Admin";
                loanDetails.InsertedDate = DateTime.Now;
                _dbContext.LoanInterestDetails.Add(loanDetails);
                _dbContext.SaveChanges();
                List<SMSDetails> sMSDetails = new List<SMSDetails>();
                var customersList = GetCustomersList();
                var text = $"Fineasee:Dear Customer,{loanDetails.BankName} bank is providing Loans For {loanDetails.LoanType} at interest rates starting at {loanDetails.MinInterestRate} and is available on our website";
                foreach (var customer in customersList)
                {
                    new LoginRepository(_dbContext).SendEmailToUser(customer.EmailId, text, false);
                    sMSDetails.Add(new SMSDetails()
                    {
                        phone_number = customer.PhoneNo,
                        text_message = text
                    });
                }
                new LoginRepository(_dbContext).SendOTPAsync(sMSDetails);
                return true;
            }

            return false;
        }
        public bool UpdateLoanDetails(LoanDetails loanInterestDetails)
        {

            var loanDetails = _dbContext.LoanInterestDetails.Where(s => s.LoanInterestDetailID == loanInterestDetails.LoanInterestDetailID).FirstOrDefault();
            if (loanDetails != null)
            {
                loanDetails.LoanType = loanInterestDetails.LoanType;
                loanDetails.BankName = loanInterestDetails.BankName;
                loanDetails.MaxloanAmount = loanInterestDetails.MaxloanAmount;
                loanDetails.MinLoanAmount = loanInterestDetails.MinLoanAmount;
                loanDetails.MaxInterestRate = loanInterestDetails.MaxInterestRate;
                loanDetails.MinInterestRate = loanInterestDetails.MinInterestRate;
                loanDetails.UpdatedDate = DateTime.Now;
                loanDetails.UpdatedBy = "Admin";
                _dbContext.SaveChanges();
                return true;
            }
            return false;
        }
        public List<string> GetLoantypes(string bankName)
        {
            var loanTypesList = _dbContext.LoanInterestDetails.Where(s => s.BankName.Contains(bankName)).Select(e=>e.LoanType).ToList();
            return loanTypesList;
        }

        public static string ReplaceUrlsWithLinks(string text)
        {
            string pattern = @"(?<=(^|\s))(https?://\S+)";
            Regex regex = new Regex(pattern);

            string replacedText = regex.Replace(text, match =>
            {
                string url = match.Value;
                return $"<a class=\"btn btn-link\" href=\"{url}\">{url}</a>";
            });

            return replacedText;
        }
        public bool CreateNewAdmin(AdminRegistrationVM adminRegistrationVM)
        {
            AdminLogin adminLogin = new AdminLogin();
            Admin admin = new Admin();
            if (adminRegistrationVM != null)
            {
                adminLogin.UserName = adminRegistrationVM.UserName;
                adminLogin.Password = adminRegistrationVM.Password;
                admin.EmailId= adminRegistrationVM.EmailId;
                admin.PhoneNo = adminRegistrationVM.PhoneNo;
                admin.FirstName = adminRegistrationVM.FirstName;
                admin.LastName = adminRegistrationVM.LastName;
                admin.InsertedBy = "Admin";
                admin.InsertedDate = DateTime.Now;
                _dbContext.Admin.Add(admin);
                _dbContext.SaveChanges();
                adminLogin.AdminId = admin.AdminId;
                adminLogin.InsertedBy = "Admin";
                adminLogin.InsertedDate = DateTime.Now;
                _dbContext.AdminLogin.Add(adminLogin);
                _dbContext.SaveChanges();
                return true;
            }

            return false;
        }
    }
}

File: schemes-python-backend/app/repositories/customer_repository.py
from sqlalchemy.orm import Session
from sqlalchemy import and_, desc
from typing import List, Optional
from datetime import datetime
import re
from app.models import (
    SchemesDetails, MultilingualSchemesData, Customer, 
    LoanInterestDetails, Admin, AdminLogin
)
from app.schemas import (
    SchemeDetailsCreate, SchemeDetailsUpdate, Customer as CustomerSchema,
    LoanDetailsCreate, LoanDetails as LoanDetailsSchema,
    AdminCreate
)

class CustomerRepository:
    def __init__(self, db: Session):
        self.db = db
    
    def get_schemes_list(self, name: str) -> List[dict]:
        """Get schemes list filtered by available_for containing name"""
        schemes = self.db.query(SchemesDetails).filter(
            and_(
                SchemesDetails.available_for.contains(name),
                SchemesDetails.is_active == True
            )
        ).order_by(desc(SchemesDetails.inserted_date)).all()
        
        return self._convert_schemes_to_dict(schemes)
    
    def get_schemes_list_by_lang_code(self, name: str, lang_code: str) -> List[dict]:
        """Get schemes list by language code"""
        multilingual_schemes = self.db.query(MultilingualSchemesData).filter(
            and_(
                MultilingualSchemesData.available_for.contains(name),
                MultilingualSchemesData.lang_code == lang_code,
                MultilingualSchemesData.is_active == True
            )
        ).all()
        
        if multilingual_schemes:
            result = []
            for scheme in multilingual_schemes:
                scheme_dict = {
                    "schemes_detail_id": scheme.schemes_detail_id,
                    "area": scheme.area,
                    "apply_and_link": scheme.apply_and_link,
                    "name_of_the_scheme": scheme.name_of_the_scheme,
                    "benefits": scheme.benefits,
                    "eligibility_criteria": scheme.eligibility_criteria,
                    "description": scheme.description,
                    "documents_required": scheme.documents_required,
                    "available_for": scheme.available_for,
                    "lang": scheme.lang_code
                }
                result.append(scheme_dict)
            return result
        else:
            # Fallback to default schemes
            return self.get_schemes_list(name)
    
    def get_all_schemes_list(self) -> List[dict]:
        """Get all active schemes"""
        schemes = self.db.query(SchemesDetails).filter(
            SchemesDetails.is_active == True
        ).order_by(desc(SchemesDetails.inserted_date)).all()
        
        return self._convert_schemes_to_dict(schemes)
    
    def add_new_scheme(self, scheme: SchemeDetailsCreate) -> int:
        """Add new scheme and return scheme ID"""
        scheme_details = SchemesDetails(
            area=scheme.area,
            apply_and_link=self._replace_urls_with_links(scheme.apply_and_link) if scheme.apply_and_link else None,
            name_of_the_scheme=scheme.name_of_the_scheme,
            benefits=scheme.benefits,
            eligibility_criteria=scheme.eligibility_criteria,
            description=scheme.description,
            documents_required=scheme.documents_required,
            available_for=scheme.available_for,
            is_active=True,
            inserted_date=datetime.utcnow(),
            inserted_by="Admin"
        )
        
        self.db.add(scheme_details)
        self.db.commit()
        self.db.refresh(scheme_details)
        
        # Add multilingual data for English
        self.add_new_scheme_by_lang_code(scheme, scheme_details.schemes_detail_id, "en")
        
        return scheme_details.schemes_detail_id
    
    def add_new_scheme_by_lang_code(self, scheme: SchemeDetailsCreate, scheme_id: int, lang_code: str) -> bool:
        """Add multilingual scheme data"""
        multilingual_scheme = MultilingualSchemesData(
            area=scheme.area,
            apply_and_link=self._replace_urls_with_links(scheme.apply_and_link) if scheme.apply_and_link else None,
            name_of_the_scheme=scheme.name_of_the_scheme,
            benefits=scheme.benefits,
            eligibility_criteria=scheme.eligibility_criteria,
            description=scheme.description,
            documents_required=scheme.documents_required,
            available_for=scheme.available_for,
            lang_code=lang_code,
            schemes_detail_id=scheme_id,
            is_active=True,
            inserted_date=datetime.utcnow(),
            inserted_by="Admin"
        )
        
        self.db.add(multilingual_scheme)
        self.db.commit()
        return True
    
    def update_scheme(self, scheme_id: int, scheme_update: SchemeDetailsUpdate) -> bool:
        """Update existing scheme"""
        scheme = self.db.query(SchemesDetails).filter(
            SchemesDetails.schemes_detail_id == scheme_id
        ).first()
        
        if scheme:
            if scheme_update.area is not None:
                scheme.area = scheme_update.area
            if scheme_update.apply_and_link is not None:
                scheme.apply_and_link = self._replace_urls_with_links(scheme_update.apply_and_link)
            if scheme_update.name_of_the_scheme is not None:
                scheme.name_of_the_scheme = scheme_update.name_of_the_scheme
            if scheme_update.benefits is not None:
                scheme.benefits = scheme_update.benefits
            if scheme_update.eligibility_criteria is not None:
                scheme.eligibility_criteria = scheme_update.eligibility_criteria
            if scheme_update.description is not None:
                scheme.description = scheme_update.description
            if scheme_update.documents_required is not None:
                scheme.documents_required = scheme_update.documents_required
            if scheme_update.available_for is not None:
                scheme.available_for = scheme_update.available_for
            if scheme_update.is_active is not None:
                scheme.is_active = scheme_update.is_active
            
            scheme.updated_date = datetime.utcnow()
            scheme.updated_by = "Admin"
            self.db.commit()
            return True
        return False
    
    def update_scheme_status(self, scheme_id: int) -> bool:
        """Deactivate scheme (soft delete)"""
        scheme = self.db.query(SchemesDetails).filter(
            SchemesDetails.schemes_detail_id == scheme_id
        ).first()
        
        if scheme:
            scheme.is_active = False
            scheme.updated_date = datetime.utcnow()
            scheme.updated_by = "Admin"
            self.db.commit()
            return True
        return False
    
    def get_customers_list(self) -> List[CustomerSchema]:
        """Get all customers"""
        customers = self.db.query(Customer).all()
        return [CustomerSchema.from_orm(customer) for customer in customers]
    
    def update_customer_status(self, customer_id: int) -> bool:
        """Toggle customer status"""
        customer = self.db.query(Customer).filter(
            Customer.customer_id == customer_id
        ).first()
        
        if customer:
            customer.customer_status = 20 if customer.customer_status == 10 else 10  # Toggle Active/Inactive
            customer.updated_date = datetime.utcnow()
            customer.updated_by = "Admin"
            self.db.commit()
            return True
        return False
    
    def get_schemes_details(self, scheme_id: int) -> Optional[dict]:
        """Get specific scheme details"""
        scheme = self.db.query(SchemesDetails).filter(
            SchemesDetails.schemes_detail_id == scheme_id
        ).first()
        
        if scheme:
            return {
                "schemes_detail_id": scheme.schemes_detail_id,
                "area": scheme.area,
                "apply_and_link": scheme.apply_and_link,
                "name_of_the_scheme": scheme.name_of_the_scheme,
                "benefits": scheme.benefits,
                "eligibility_criteria": scheme.eligibility_criteria,
                "description": scheme.description,
                "documents_required": scheme.documents_required,
                "available_for": scheme.available_for
            }
        return None
    
    def get_loan_details(self, bank_name: str, loan_type: str) -> Optional[LoanDetailsSchema]:
        """Get loan details by bank name and loan type"""
        loan = self.db.query(LoanInterestDetails).filter(
            and_(
                LoanInterestDetails.bank_name.contains(bank_name),
                LoanInterestDetails.loan_type == loan_type
            )
        ).first()
        
        if loan:
            return LoanDetailsSchema.from_orm(loan)
        return None
    
    def get_loan_types(self, bank_name: str) -> List[str]:
        """Get loan types for a specific bank"""
        loan_types = self.db.query(LoanInterestDetails.loan_type).filter(
            LoanInterestDetails.bank_name.contains(bank_name)
        ).distinct().all()
        
        return [loan_type[0] for loan_type in loan_types]
    
    def get_loan_details_list(self) -> List[LoanDetailsSchema]:
        """Get all loan details"""
        loans = self.db.query(LoanInterestDetails).all()
        return [LoanDetailsSchema.from_orm(loan) for loan in loans]
    
    def add_loan_details(self, loan_details: LoanDetailsCreate) -> bool:
        """Add new loan details"""
        loan = LoanInterestDetails(
            bank_name=loan_details.bank_name,
            loan_type=loan_details.loan_type,
            min_loan_amount=loan_details.min_loan_amount,
            max_loan_amount=loan_details.max_loan_amount,
            min_interest_rate=loan_details.min_interest_rate,
            max_interest_rate=loan_details.max_interest_rate,
            inserted_by="Admin",
            inserted_date=datetime.utcnow()
        )
        
        self.db.add(loan)
        self.db.commit()
        
        # Send notifications to customers (placeholder)
        customers = self.get_customers_list()
        message = f"New loan offer: {loan_details.bank_name} bank is providing {loan_details.loan_type} loans at interest rates starting from {loan_details.min_interest_rate}%"
        
        # Here you would integrate with notification service
        print(f"Sending notifications to {len(customers)} customers: {message}")
        
        return True
    
    def update_loan_details(self, loan_id: int, loan_details: LoanDetailsCreate) -> bool:
        """Update existing loan details"""
        loan = self.db.query(LoanInterestDetails).filter(
            LoanInterestDetails.loan_interest_detail_id == loan_id
        ).first()
        
        if loan:
            loan.bank_name = loan_details.bank_name
            loan.loan_type = loan_details.loan_type
            loan.min_loan_amount = loan_details.min_loan_amount
            loan.max_loan_amount = loan_details.max_loan_amount
            loan.min_interest_rate = loan_details.min_interest_rate
            loan.max_interest_rate = loan_details.max_interest_rate
            loan.updated_date = datetime.utcnow()
            loan.updated_by = "Admin"
            self.db.commit()
            return True
        return False
    
    def create_new_admin(self, admin_data: AdminCreate) -> bool:
        """Create new admin"""
        # Create Admin record
        admin = Admin(
            first_name=admin_data.first_name,
            last_name=admin_data.last_name,
            email_id=admin_data.email_id,
            phone_no=admin_data.phone_no,
            inserted_by="SuperAdmin",
            inserted_date=datetime.utcnow()
        )
        
        self.db.add(admin)
        self.db.commit()
        self.db.refresh(admin)
        
        # Create AdminLogin record
        from repositories.login_repository import LoginRepository
        login_repo = LoginRepository(self.db)
        hashed_password = login_repo._hash_password(admin_data.password)
        
        admin_login = AdminLogin(
            admin_id=admin.admin_id,
            email_id=admin_data.email_id,
            user_name=admin_data.username,
            password=hashed_password,
            login_status=False,
            inserted_by="SuperAdmin",
            inserted_date=datetime.utcnow()
        )
        
        self.db.add(admin_login)
        self.db.commit()
        return True
    
    def _convert_schemes_to_dict(self, schemes: List[SchemesDetails]) -> List[dict]:
        """Convert SQLAlchemy models to dictionaries"""
        result = []
        for scheme in schemes:
            scheme_dict = {
                "schemes_detail_id": scheme.schemes_detail_id,
                "area": scheme.area,
                "apply_and_link": scheme.apply_and_link,
                "name_of_the_scheme": scheme.name_of_the_scheme,
                "benefits": scheme.benefits,
                "eligibility_criteria": scheme.eligibility_criteria,
                "description": scheme.description,
                "documents_required": scheme.documents_required,
                "available_for": scheme.available_for,
                "is_active": scheme.is_active
            }
            result.append(scheme_dict)
        return result
    
    def _replace_urls_with_links(self, text: str) -> str:
        """Replace URLs with HTML links"""
        if not text:
            return text
        
        url_pattern = r'(?<=(^|\s))(https?://\S+)'
        
        def replace_url(match):
            url = match.group(0)
            return f'<a class="btn btn-link" href="{url}">{url}</a>'
        
        return re.sub(url_pattern, replace_url, text)
